src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4146,7 +4146,7 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
             TypeWithState rightResult = VisitOptionalImplicitConversion(rightOperand, targetType, useLegacyWarnings: UseLegacyWarnings(leftOperand, targetType), trackMembers: false, AssignmentKind.Assignment);
             TrackNullableStateForAssignment(rightOperand, targetType, leftSlot, rightResult, MakeSlot(rightOperand));
             Join(ref this.State, ref leftState);
-            TypeWithState resultType = GetNullCoalescingResultType(rightResult, targetType.Type);
+            TypeWithState resultType = GetNullCoalescingResultType(rightResult, TypeWithState.Create(targetType.Type, NullableFlowState.NotNull));
```

Comment:
I think this is just doing `TypeWithState.Create(targetType.Type, rightResult.State)` which seems simpler.